play-route/src/main/scala/org/http4s/server/play/PlayRouteBuilder.scala

```diff
@@ -0,0 +1,149 @@
+            Header(headerName, value)
+          }
+      }),
+      body = EmptyBody
+    )
+
+  type ResponseStream = fs2.Stream[F, Byte]
+
+  def convertStream(responseStream: ResponseStream): PlayTargetStream = {
+    val entityBody: fs2.Stream[F, Byte] = responseStream
+    Source
+      .fromPublisher(entityBody.toUnicastPublisher())
+      .map(byte => ByteString(byte))
+  }
+
+  /**
+    * A Play accumulator Sinks HTTP data in, and then pumps out a future of a Result.
+    * That Result will have a Source as the response HTTP Entity.
+    *
+    * Here we create a unattached sink, map its materialized value into a publisher,
+    * convert that into an FS2 Stream, then pipe the request body into the http4s request.
+    */
+  def playRequestToPlayResponse(requestHeader: RequestHeader): PlayAccumulator = {
+    val sink: Sink[ByteString, Future[Result]] = {
+      Sink.asPublisher[ByteString](false).mapMaterializedValue { publisher =>
+        val requestBodyStream: fs2.Stream[F, Byte] =
+          publisher.toStream().flatMap(bs => fs2.Stream.fromIterator[F, Byte](bs.toIterator))
+
+        val wrappedResponse: F[Response[F]] =
+          F.map(unwrappedRun(
+            requestHeaderToRequest(requestHeader).withBodyStream(requestBodyStream)).value)(_.get)
+        val wrappedResult: F[Result] = F.map(wrappedResponse) { response =>
+          Result(
+            header = convertResponseToHeader(response),
+            body = Streamed(
+              data = convertStream(response.body),
+              contentLength = response.contentLength,
+              contentType = response.contentType.map(_.value)
+            )
+          )
+        }
+
+        val promise = Promise[Result]
+
+        F.runAsync(wrappedResult) {
+            case Left(bad) =>
+              promise.failure(bad)
+              IO.unit
+            case Right(good) =>
+              promise.success(good)
+              IO.unit
+          }
+          .unsafeRunSync()
+
+        promise.future
+      }
+    }
+    Accumulator.apply(sink)
+  }
+
+  def convertResponseToHeader(response: Response[F]): ResponseHeader =
+    ResponseHeader(
+      status = response.status.code,
+      headers = response.headers.collect {
+        case header
+            if !PlayRouteBuilder.AkkaHttpSetsSeparately.contains(header.parsed.name.value) =>
+          header.parsed.name.value -> header.parsed.value
+      }.toMap
+    )
+
+  /** Big big hack :-( Teach me! **/
+  def routeMatches(requestHeader: RequestHeader): Boolean = {
+    val optionalResponse: OptionT[F, Response[F]] =
+      unwrappedRun.apply(requestHeaderToRequest(requestHeader))
+    val efff: F[Option[Response[F]]] = optionalResponse.value
+    val completion = Promise[Boolean]()
+    F.runAsync(efff) {
+        case Left(f) => completion.failure(f); IO.unit
+        case Right(s) => completion.success(s.isDefined); IO.unit
+      }
+      .unsafeRunSync()
+    Await.result(completion.future, Duration.Inf)
```

Comment:
Does this have to return a `Boolean` instead of a `Future[Boolean]`?  This is going to block not just while the routing is done, but also while the response is generated.  If you can tie it into Play as a `Future`, life is going to be a lot beter.

Reply:
Unfortunately the route-checking in Play is synchronous (route matcher is a `PartialFunction`, so `.isDefinedAt` is called synchronously). I wonder if there's a better way to do this, cause I really don't like it much myself at all.

Reply:
The cleanest thing you can do, would be to be able to introspect your route (like in [rho](https://github.com/http4s/rho)) to the extract the endpoints into the routes file play wants, and separate each action endpoint into a play action.

The problem is play routes matching is essentially not directly compatible with http4s without a hacky approach you plug into their routes. That is: Checking routes if they match http4s' routes first, then running the rest. This is not going to be fun to do, at all.

You are better off building a sort of structure from the routes (with `Request[F] => F[Response[F]] => Future[PlayResponse]`), then turning each one into a play action, instead of the whole batch into it.